File: examples/governance/compliance_domain/lib/compliance_domain/governance_policy/commands/retire_policy.rb
module ComplianceDomain
  class GovernancePolicy
    module Commands
      class RetirePolicy
        include Hecks::Command
        emits "RetiredPolicy"

        attr_reader :policy_id

        def initialize(policy_id: nil)
          @policy_id = policy_id
        end

        def call
          existing = repository.find(policy_id)
          if existing
            unless ["active", "suspended"].include?(existing.status)
              raise Hecks::Error, "Cannot RetirePolicy: status must be one of active, suspended, got '#{existing.status}'"
            end
            GovernancePolicy.new(
              id: existing.id,
              name: existing.name,
              description: existing.description,
              category: existing.category,
              framework_id: existing.framework_id,
              effective_date: existing.effective_date,
              review_date: existing.review_date,
              requirements: existing.requirements,
              status: "retired"
            )
          else
            raise Hecks::Error, "GovernancePolicy not found: #{policy_id}"
          end
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/retire_policy.go
package domain

import (
	"time"
	"fmt"
)

type RetirePolicy struct {
	PolicyId string `json:"policy_id"`
}

func (c RetirePolicy) CommandName() string { return "RetirePolicy" }

func (c RetirePolicy) Execute(repo GovernancePolicyRepository) (*GovernancePolicy, *RetiredPolicy, error) {
	existing, err := repo.Find(c.PolicyId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("GovernancePolicy not found: %s", c.PolicyId)
	}
	if existing.Status != "active" && existing.Status != "suspended" {
		return nil, nil, fmt.Errorf("cannot RetirePolicy: GovernancePolicy is in %s state", existing.Status)
	}
	existing.Status = "retired"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := RetiredPolicy{
		AggregateID: existing.ID,
		PolicyId: c.PolicyId,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}
